server/src/main/java/org/elasticsearch/search/crossproject/CrossProjectIndexResolutionValidator.java

```diff
@@ -103,82 +111,143 @@ public static ElasticsearchException validate(
+                        }
                     }
                 }
             } else {
-                ElasticsearchException localException = checkResolutionFailure(
-                    localExpressions,
-                    result,
-                    originalExpression,
-                    indicesOptions
-                );
                 if (localException == null && localExpressions != ResolvedIndexExpression.LocalExpressions.NONE) {
                     // found locally, continue to next expression
                     continue;
                 }
-                ElasticsearchSecurityException unauthorizedException = null;
-                if (localException instanceof ElasticsearchSecurityException securityException) {
-                    unauthorizedException = securityException;
-                }
                 assert localExpressions != ResolvedIndexExpression.LocalExpressions.NONE || false == remoteExpressions.isEmpty()
                     : "both local expression and remote expressions are empty which should have errored earlier at index rewriting time";
+                ElasticsearchSecurityException localSecurityException = null;
+                if (localException instanceof ElasticsearchSecurityException securityException) localSecurityException = securityException;
+
                 boolean foundFlat = false;
                 // checking if flat expression matched remotely
                 for (String remoteExpression : remoteExpressions) {
                     String[] splitResource = splitQualifiedResource(remoteExpression);
-                    ElasticsearchException exception = checkSingleRemoteExpression(
+                    var projectAlias = splitResource[0];
+                    var resource = splitResource[1];
+
+                    ElasticsearchException remoteException = checkSingleRemoteExpression(
                         remoteResolvedExpressions,
-                        splitResource[0], // projectAlias
-                        splitResource[1], // resource
+                        projectAlias,
+                        resource,
                         remoteExpression,
                         indicesOptions
                     );
-                    if (exception == null) {
+                    if (remoteException == null) {
                         // found flat expression somewhere
                         foundFlat = true;
+                        // ensure we don't report this index as unauthorized if it's unauthorized on another remote
+                        if (remoteUnauthorizedIndices != null) {
+                            for (var entry : remoteUnauthorizedIndices.entrySet()) {
+                                entry.setValue(entry.getValue().stream().filter(index -> index.endsWith(":" + resource) == false).toList());
+                            }
+                        }
```

Comment:
We still need to clear `remoteAuthorizationExceptions`. Instead of clearing, I'd prefer we don't add to the maps until we are sure the expression is not found anywhere.